### PR TITLE
Another fake PR for review (do not close!) 

### PR DESCRIPTION
**Pull request by egparedes**

### CODE REVIEW

File: src/py2spack/__init__.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+"""py2spack package.
+
+This package allows converting standard python packages to Spack packages.
+"""
```
Comment (egparedes) [inline src/py2spack/__init__.py]: The first line should be the summary, not  the name of the package (which is already encoded in the folder or file name). 
```suggestion
"""A package to convert Python distribution packages to Spack packages."""
```

All comments like this one about general conventions or style rules apply to all files in the repo, even if I will only the mention it the first time they appear.

File: src/py2spack/__init__.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+from .main import SpackPyPkg
+
+__version__ = "0.0.1"
+
+__all__ = ["SpackPyPkg"]
```
Comment (egparedes) [inline src/py2spack/__init__.py]: Don't forget that our coding conventions (Google coding conventions in this case) build on top of PEP 8, the most basic language style convention which says that the special/magic/dunder module variables should be defined before any regular import:    https://peps.python.org/pep-0008/#module-level-dunder-names
```suggestion
__all__ = ["SpackPyPkg"]
__version__ = "0.0.1"

from .main import SpackPyPkg
```
The rationale for this is to set the module metadata before anything else, to avoid possible errors when importing code relying in this metadata being already defined.

File: src/py2spack/loading.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+class APIError(Exception):
+    """Error during version lookup or querying of the PyPI JSON API."""
+
+    def __init__(
+        self,
+        msg: str,
+    ):
+        """Initialize error."""
+        super().__init__(msg)
```
Comment (egparedes) [inline src/py2spack/loading.py]: The current `__init__` is completely redundant since it's only calling the `__init__` of the superclass. Additionally, the name of the class is too generic and should be more specific, e.g.   
```suggestion
class PyPIQueryError(Exception):   # Or QueryError, or something along these lines
    """Error during version lookup or querying of the PyPI JSON API."""
    pass
    
```

File: src/py2spack/loading.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+
+def _acceptable_version(version: str) -> Optional[vn.Version]:
+    """Check whether version string can be parsed and has correct format."""
```
Comment (egparedes) [inline src/py2spack/loading.py]: Either this function is a predicate checking if the argument is a valid version, in which case it should be named something like `is_acceptable_version()` and return a bool, or is a function parsing version strings which returns `None` if there are errors, and then it should be named like `parse_version()` or `parse_supported_version()`. This case looks like the latter, so I'd suggest to rename it and update the docstring.

In general, functions should be named using verbs, and if they return booleans, they should be named as predicate expressions (`is_larger_than()`, `starts_with()`, `has_children()` , ...)

File: src/py2spack/loading.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
+"""Utils for downloading packages from PyPI and opening the pyproject.tomls."""
+
+import io
+import tarfile
+from typing import Any, Dict, List, Optional
```
Comment (egparedes) [inline src/py2spack/loading.py]: From python 3.9 builtin types like `dict` and `list` are generic and therefore they should be used directly, since their aliases in typing are deprecated (https://typing.readthedocs.io/en/latest/guides/modernizing.html#generics-in-the-typing-module)
```suggestion
from typing import Any, Optional
```

File: src/py2spack/loading.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+def _acceptable_version(version: str) -> Optional[vn.Version]:
+    """Check whether version string can be parsed and has correct format."""
+    try:
+        v = vn.parse(version)
+        # do not support post releases of prereleases etc.
+        if v.pre and (v.post or v.dev or v.local):
+            return None
+        return v
+    except vn.InvalidVersion:
+        return None
```
Comment (egparedes) [inline src/py2spack/loading.py]: In general the control flow of the code should be as linear as possible to make it easier to understand:  
```suggestion
    result = None
    try:
        v = vn.parse(version)
        # do not support post releases of prereleases etc.
        if not (v.pre and (v.post or v.dev or v.local)):
            result = v
    except vn.InvalidVersion:
        pass

    return result
```

File: src/py2spack/loading.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+
+class PyPILookup:
```
Comment (egparedes) [inline src/py2spack/loading.py]: This class looks like the specific implementation using PyPI API of a more general concept of some "provider" of python distribution packages. Wouldn't it make more sense to define a `Protocol` for the general provider interface and then a concrete implementation of the interface connecting to PyPI? In this way, different providers (e.g. github?) can be used transparently without changing other components.

Basic sketch of the proposal:
```python
class PyProjectProvider(Protocol):

    @abc.abstractmethod
    def get_versions(self, name: str) -> list[vn.Version]:
        ...

    @abc.abstractmethod
    def get_files(
        self, name: str, versions: int | list[vn.Version] | None  = None
    ) -> List[Dict[Any, Any]] | Error:
         ....

class PyPIAPIProvider(DistributionPackagesProvider):
     url: str =  "https://pypi.org/simple/"
     ....

```
Also, I'm not completely sure if `get_files` is the best interface for this. Wouldn't it be cleaner to have public a method to get the `pyproject` files for some version and another to get the actual distributions packages? Of course both could be implemented in terms of a `_get_files()` helper method

Finally, consider the idea of using a parameter for the URL of the API in the concrete PyPI implementation, so the same class is potentially reusable for different PyPI deployments...

File: src/py2spack/loading.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+
+    def get_versions(self, name: str) -> List[vn.Version]:
+        """Get acceptable versions for package (from cache if possible)."""
+        if name not in self.version_cache:
+            data = self._get(name)
+
+            if isinstance(data, APIError):
+                self.version_cache[name] = []
+
+        return self.version_cache[name]
```
Comment (egparedes) [inline src/py2spack/loading.py]: The `fucntools` module comes with memoization functionality (`cache` or `lru_cache`) that could be leveraged here to simplify the implementation. Also I would cleanly decouple the handling of the query to the server and the processing of the returned information. For example:

```suggestion
    def _get_data(self, name: str) -> dict | APIError:
        """Load info for the available distribution files from PyPI."""
        r = requests.get(
            f"https://pypi.org/simple/{name}/",
            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
            timeout=10,
        )
        if r.status_code != 200:
            if r.status_code == 404:
                msg = f"Package {name} not found on PyPI (status code 404)"
                return APIError(msg)

            msg = (
                f"Error when querying JSON API (status code {r.status_code})."
                f" Response: {r.text}"
            )
            return APIError(msg)

        data: dict = r.json()
        return data

    @functools.lru_cache
    def get_versions(self, name: str) -> List[vn.Version]:
        """Get acceptable versions for package (from cache if possible)."""
        data = self._get(name)
        if isinstance(data, APIError):   ## I think this behavior is wrong and it should return an error in this case
            self.version_cache[name] = []
        versions = data["versions"]

        # parse and sort versions
        parsed_versions = sorted(
            {vv for v in versions if (vv := _acceptable_version(v))}
        )

        return parsed_versions
```

File: src/py2spack/loading.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+    def get_files(
```
Comment (egparedes) [inline src/py2spack/loading.py]: See related comments above.

File: src/py2spack/loading.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        if isinstance(data, APIError):
+            return data
+        files = data["files"]
+
+        # for now we only support tarball archives like .tar.gz
+        files_known_format = [
+            f for f in files if _known_archive_format(f["filename"])
+        ]
+
+        if len(files_known_format) == 0:
```
Comment (egparedes) [inline src/py2spack/loading.py]: The idiomatic Python way of checking if builtin collections are empty is check if they are false values:
```suggestion
        if not files_known_format:
```
Also: https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations

File: src/py2spack/loading.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+            if isinstance(archive_ext, APIError):
+                continue
+
+            v = _parse_version(filename, name, archive_ext)
+
+            if v is None:
+                continue
+
+            try:
+                sha256 = f["hashes"]["sha256"]
```
Comment (egparedes) [inline src/py2spack/loading.py]: Can you explain to me when the "sha256" hash could be missing? Why is it correct to continue in that case?

File: src/py2spack/loading.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+
+        return files_parsed[-last_n_versions:]
+
+
+def _parse_archive_extension(filename: str) -> str | APIError:
+    extension_list = [
+        ext for ext in KNOWN_ARCHIVE_FORMATS if filename.endswith(ext)
+    ]
+
+    if len(extension_list) == 0:
```
Comment (egparedes) [inline src/py2spack/loading.py]:
```suggestion
    if not extension_list:
```

File: src/py2spack/loading.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        # we return an API error here because the filenames are obtained through
+        # the API and the function is used during the API lookup process
+        msg = f"Extension not recognized for: {filename}"
+        return APIError(msg)
+
+    if len(extension_list) == 1:
+        return extension_list[0]
```
Comment (egparedes) [inline src/py2spack/loading.py]: This case is redundant, since the default case below will return the same value.

File: src/py2spack/loading.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+    # get the longest matching extension, e.g. .tar.gz instead of .gz
+    longest_matching_ext = max(extension_list, key=len)
+    return longest_matching_ext
+
+
+def _parse_version(
+    filename: str, pkg_name: str, archive_ext: str
+) -> vn.Version | None:
+    """Parse version from filename and check correct formatting."""
+    prefix = pkg_name + "-"
```
Comment (egparedes) [inline src/py2spack/loading.py]:
```suggestion
    prefix = f"{pkg_name}-"
```

File: src/py2spack/loading.py [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+    try:
+        parsed_version: vn.Version = vn.parse(version_str)
+        return parsed_version
+
+    except vn.InvalidVersion:
+        return None
+
+
+def _known_archive_format(filename: str):
+    return any([filename.endswith(ext) for ext in KNOWN_ARCHIVE_FORMATS])
```
Comment (egparedes) [inline src/py2spack/loading.py]: There is no need to allocate a list for the result of the comprehensions since it's just a temporary value.
```suggestion
    return any(filename.endswith(ext) for ext in KNOWN_ARCHIVE_FORMATS)
```

File: src/py2spack/loading.py [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+
+def _download_sdist(url: str) -> io.BytesIO | APIError:
```
Comment (egparedes) [inline src/py2spack/loading.py]: This function looks like a general utility without specific business logic. I will move it to some utility module (e.g. `utils.py`)

File: src/py2spack/loading.py [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+    else:
+        filename = url.split("/")[-1]
+        msg = (
+            f"Failed to download sdist {filename},"
+            f" status code: {response.status_code}"
+        )
+        return APIError(msg)
+
+
+def _extract_from_tar(
```
Comment (egparedes) [inline src/py2spack/loading.py]: This function looks like a general utility without specific business logic. I will move it to some utility module (e.g. `utils.py`)

File: src/py2spack/loading.py
Comment (egparedes) [inline src/py2spack/loading.py:1]: I think the name of the module is not very precise. What do you think of something along the lines of  `package_providers`, `project_providers`... ?

**Review by egparedes — COMMENTED** [review]
Another part of the review: this times I only reviewed the `loading.py` module, although some of the comments are more general and also apply to other modules. Feel free to add your questions and objections here so we can continue the discussion in this fake PR review.

File: src/py2spack/loading.py [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+
+    def get_versions(self, name: str) -> List[vn.Version]:
+        """Get acceptable versions for package (from cache if possible)."""
+        if name not in self.version_cache:
+            data = self._get(name)
+
+            if isinstance(data, APIError):
+                self.version_cache[name] = []
+
+        return self.version_cache[name]
```
Comment (egparedes) [inline src/py2spack/loading.py]: A similar caching functionality is already provided by `functools.lru_cache()` or `functools.cache()`. Also, the parsing of the versions doesn't belong to the utility method contacting the server but to the actual method dealing with versions.

```suggestion
    def _get(self, name: str) -> dict | APIError:
        """Load info for the available distribution files from PyPI.

        Available versions are cached.
        """
        r = requests.get(
            f"https://pypi.org/simple/{name}/",
            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
            timeout=10,
        )
        if r.status_code != 200:
            if r.status_code == 404:
                msg = f"Package {name} not found on PyPI (status code 404)"
                return APIError(msg)

            msg = (
                f"Error when querying JSON API (status code {r.status_code})."
                f" Response: {r.text}"
            )
            return APIError(msg)

        data: dict = r.json()
        return data

    @functools.cache
    def get_versions(self, name: str) -> List[vn.Version]:
        """Get acceptable versions for package (from cache if possible)."""
        data = self._get(name) 
        if isinstance(data, APIError):
            return []   ## <----- I don't think it's right to return an empty list instead of an error here...
        versions = data["versions"]

        # parse and sort versions
        parsed_versions = sorted(
            {vv for v in versions if (vv := _acceptable_version(v))}
        )
        return parsed_versions
```

File: src/py2spack/loading.py (hunk 7 above)
Comment (davhofer) [reply src/py2spack/loading.py]: @egparedes Regarding the `get_files` interface: 
The method currently only provides metadata of the distribution package like filename, download url, hash, etc.
But it would make sense to directly put methods for loading the pyproject.toml file contents etc. here too. And I agree, this could be done for a specific version of the package. 
I think a public method to get the `pyproject` should be enough though (no public method to get the actual distribution package), since we only need the distribution package to get the pyproject.toml data. So this can just be downloaded and extracted in a helper method.

So in summary, my idea would be to do a protocol with public methods `get_versions(name)` which returns a list of available versions (already filtered for only source distributions?) and `get_pyproject(name, version)` which returns a dict with the contents of the pyproject.toml file.

If later we also need to get other files from the source distribution except the `pyproject.toml`, I could also make a more generic method to extract a file from the source distribution?

File: src/py2spack/loading.py (hunk 11 above)
Comment (davhofer) [reply src/py2spack/loading.py]: @egparedes While the `hashes` field is required, it can be that there is no hash in it, or another hash than `sha256`, according to https://peps.python.org/pep-0691/.
In that case I want to simply skip this specific version and try the next one.

Comment (davhofer) [reply src/py2spack/loading.py]: spack actually also supports other hashes than sha256, so I could in addition also check for those

File: src/py2spack/loading.py (hunk 7 above)
Comment (egparedes) [reply src/py2spack/loading.py]: Yes, I agree. Let's keep it simple for now.
